bindings/java/src/org/sleuthkit/datamodel/TskUnsupportedSchemaVersionException.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * SleuthKit Java Bindings
+ *
+ * Copyright 2011-2017 Basis Technology Corp.
+ * Contact: carrier <at> sleuthkit <dot> org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.sleuthkit.datamodel;
+
+/**
+ * Subtype of TskCoreException that is thrown when attempting to open a database
+ * with an unsupported schema version.
+ */
+public class TskUnsupportedSchemaVersionException extends TskCoreException {
+
+	private static final long serialVersionUID = 1L;
+	private final DBSchemaVersion unsupportedVersion;
+	private final DBSchemaVersion currentVersion;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param unsupportedVer The schema version of the db that couldn't be
+	 *                       opened.
+	 * @param currentVer     The current schema version in the code.
+	 * @param msg            A message with details.
+	 */
+	TskUnsupportedSchemaVersionException(DBSchemaVersion unsupportedVer, DBSchemaVersion currentVer, String msg) {
+		super(msg);
+		this.unsupportedVersion = unsupportedVer;
+		this.currentVersion = currentVer;
+	}
+
+	/**
+	 * Get the version of the schema used by the db that we attempted to open.
+	 *
+	 * @return The version of the schema used by the db that we attempted to
+	 *         open.
+	 */
+	public DBSchemaVersion getUnsupportedVersion() {
+		return unsupportedVersion;
+	}
+
+	/**
+	 * The current version of the schema that the code uses.
+	 *
+	 * @return The current version of the schema that the code uses.
+	 */
+	public DBSchemaVersion getCurrentVersion() {
+		return currentVersion;
+	}
+}
```

Comment:
If we are going to narrow this class by providing state (as opposed to just supporting the base class constructors), I think we would do well to use that state to provide overrides of getMessage, getLocalizedMessage (can we?), and toString. The typical client of an Exception class would be very unlikely to try to use more specialized methods, since exception handling code is usually pretty generic. So I vote we remove the getters altogether as too specific to the Autopsy use case. I am certainly comfortable with the "user-friendly" message generated for this exception in Autopsy just stating incompatibility, as I don't see users caring about the version number details. In fact, it seems we might be best served to by doing way with the "non-standard" constructor and allowing the client of the class to put in whatever message they see fit, with the name of the class conveying the general idea of the exception type to the recipient. I think that is the usual thing to do when defining Exception subclasses (e.g., see https://docs.oracle.com/javase/tutorial/essential/exceptions/creating.html, https://stackoverflow.com/questions/1754315/how-to-create-custom-exceptions-in-java, etc.).